tensorflow-framework/src/test/java/org/tensorflow/framework/utils/ND.java

```diff
@@ -647,6 +731,97 @@ public static FloatNdArray sum(FloatNdArray a, Integer[] axes, boolean keepDims)
     }
   }
 
+  /**
+   * Sum all elements of an array
+   *
+   * @param a the array
+   * @return an a array with one element containing the sum.
+   */
+  public static DoubleNdArray sum(DoubleNdArray a) {
+    AtomicReference<Double> sum = new AtomicReference<>(0.);
+    a.scalars().forEach(f -> sum.set(sum.get() + f.getDouble()));
+    return NdArrays.scalarOf(sum.get());
+  }
+
+  /**
+   * Sum all elements of an array based on the specified axis
+   *
+   * @param a the array
+   * @param axis the axis to sum
+   * @return an a array the sum over the axis less the diemsnion
+   */
+  public static DoubleNdArray sum(DoubleNdArray a, int axis) {
+    return sum(a, axis, false);
+  }
+
+  /**
+   * Sum all elements of an array based on the specified axis
+   *
+   * @param a the array
+   * @param axis the axis to sum
+   * @param keepDims indicates whether the dimensions over the sum should be kept or not.
+   * @return an a array the sum over the axis
+   */
+  public static DoubleNdArray sum(DoubleNdArray a, int axis, boolean keepDims) {
+    Shape shape = a.shape();
+    int nDims = shape.numDimensions();
+    int xis = nDims - 1 - axis;
+    long totalSize = shape.size();
+    long axisSize = shape.size(xis);
+    final double[] sums = new double[(int) axisSize];
+
+    a.scalars()
+            .forEachIndexed(
+                    (idx, f) -> {
+                      sums[(int) idx[xis]] += f.getDouble();
+                    });
+
+    if (keepDims) {
+      long[] newDims = shape.asArray();
+      newDims[axis] = 1;
+      final AtomicInteger counter = new AtomicInteger();
+      DoubleNdArray arrayK = NdArrays.ofDoubles(Shape.of(newDims));
+      arrayK
+              .elements(newDims.length - 1)
+              .forEachIndexed(
+                      (idx, v) -> {
+                        v.setDouble(sums[counter.getAndAdd(1)]);
+                      });
+      return arrayK;
+    } else {
+      return NdArrays.vectorOf(sums);
+    }
+  }
+
+  /**
+   * Sum all elements of an array based on the specified axis
+   *
+   * @param a the array
+   * @param axes the axis to sum
+   * @param keepDims indicates whether the dimensions over the sum should be kept or not.
+   * @return an a array the sum over the axis
+   */
+  public static DoubleNdArray sum(DoubleNdArray a, Integer[] axes, boolean keepDims) {
+    Shape shape = a.shape();
+    if (axes == null) {
+      DoubleNdArray result = sum(a);
+      if (keepDims) {
+        double scalar = result.getDouble(0);
+        long[] dims = {1, 1};
```

Comment:
Only supports 2d arrays.

Reply:
By design right now, I just did enough for testing.

A `NumPy` replacement based on `ndarray`, would be a nice feature to have if you want to explore more.
This was brought up at yesterday's meeting. [TensorFlow SIG JVM Notes](https://docs.google.com/document/d/1RqSIKKFLE8kMSjUuZWsVS4JNCRHIrCYiuWcFZRNk2oU/edit#heading=h.ac8mm2gqdm6b)

Reply:
Perhaps rename this method `sum2D` and validate `a` accordingly?

Reply:
👍 Fine -- it's a test class